File: app/keyboards.py
"""
app/keyboards.py
-----------------
All inline keyboard builders live here, kept separate from handlers so
layout can be tweaked without touching business logic.
"""

from __future__ import annotations

from aiogram.types import InlineKeyboardButton, InlineKeyboardMarkup

from app.commands import PLATFORM_LABELS, CiscoCommand

CB_CMD = "cmd"          # cmd:<id>                 -> open a command's detail view
CB_CATEGORY = "cat"      # cat:<name>                -> list commands in a category
CB_CATEGORIES = "cats"   # cats                      -> show category picker
CB_PLATFORM = "plat"     # plat:<id>:<platform>      -> show single-platform view
CB_CONVERT = "conv"      # conv:<id>                 -> show cross-platform table
CB_FAVORITE = "fav"      # fav:<id>                  -> toggle favorite
CB_RANDOM = "rand"       # rand                      -> new random command
CB_EXPLAIN = "expl"      # expl:<id>                 -> full explanation view
CB_BACK_RESULTS = "back" # back                      -> not used across sessions (stateless)
CB_NOOP = "noop"


def command_detail_keyboard(cmd: CiscoCommand, is_favorite: bool) -> InlineKeyboardMarkup:
    fav_label = "\u2b50 Remove Favorite" if is_favorite else "\u2606 Add Favorite"
    rows = [
        [
            InlineKeyboardButton(text="\U0001f5c2 Explain", callback_data=f"{CB_EXPLAIN}:{cmd.id}"),
            InlineKeyboardButton(text="\U0001f504 Convert", callback_data=f"{CB_CONVERT}:{cmd.id}"),
        ],
        [
            InlineKeyboardButton(text=fav_label, callback_data=f"{CB_FAVORITE}:{cmd.id}"),
        ],
        [
            InlineKeyboardButton(
                text=f"\U0001f4c2 {cmd.category}", callback_data=f"{CB_CATEGORY}:{cmd.category}"
            ),
        ],
    ]
    return InlineKeyboardMarkup(inline_keyboard=rows)


def platform_picker_keyboard(cmd: CiscoCommand) -> InlineKeyboardMarkup:
    row = [
        InlineKeyboardButton(
            text=PLATFORM_LABELS[p], callback_data=f"{CB_PLATFORM}:{cmd.id}:{p}"
        )
        for p in ("ios", "ios_xe", "ios_xr", "nx_os")
    ]
    return InlineKeyboardMarkup(inline_keyboard=[row])


def categories_keyboard(categories: list[str]) -> InlineKeyboardMarkup:
    rows = []
    row: list[InlineKeyboardButton] = []
    for i, cat in enumerate(categories, start=1):
        row.append(InlineKeyboardButton(text=cat, callback_data=f"{CB_CATEGORY}:{cat}"))
        if i % 2 == 0:
            rows.append(row)
            row = []
    if row:
        rows.append(row)
    return InlineKeyboardMarkup(inline_keyboard=rows)


def search_results_keyboard(results) -> InlineKeyboardMarkup:
    rows = [
        [InlineKeyboardButton(text=f"{r.command.title}", callback_data=f"{CB_CMD}:{r.command.id}")]
        for r in results
    ]
    return InlineKeyboardMarkup(inline_keyboard=rows)


def command_list_keyboard(commands: list[CiscoCommand]) -> InlineKeyboardMarkup:
    rows = [
        [InlineKeyboardButton(text=c.title, callback_data=f"{CB_CMD}:{c.id}")]
        for c in commands
    ]
    return InlineKeyboardMarkup(inline_keyboard=rows)


def random_again_keyboard() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        inline_keyboard=[[InlineKeyboardButton(text="\U0001f3b2 Another random", callback_data=CB_RANDOM)]]
    )

File: src/keyboards.ts
import type { CiscoCommand, InlineKeyboardMarkup, Category, Platform } from "./types";
import { PLATFORM_LABELS } from "./types";
import { escapeMarkdown } from "./telegram";

/** Full detail card for a command: purpose, syntax, args, example, output, notes, etc. */
export function formatCommandDetail(cmd: CiscoCommand): string {
  const lines: string[] = [];
  lines.push(`*${escapeMarkdown(cmd.title)}*`);
  lines.push(`_Category: ${escapeMarkdown(cmd.category)}_`);
  lines.push("");

  if (cmd.description) {
    lines.push(`*Purpose:*\n${cmd.description}`);
    lines.push("");
  }

  lines.push("*Platform Support:*");
  const platforms: Platform[] = ["ios", "ios_xe", "ios_xr", "nx_os"];
  for (const p of platforms) {
    const val = cmd[p];
    lines.push(`• *${PLATFORM_LABELS[p]}:* ${val ? "\`" + val + "\`" : "_not supported / no direct equivalent_"}`);
  }
  lines.push("");

  if (cmd.syntax) {
    lines.push(`*Syntax:*\n\`${cmd.syntax}\``);
    lines.push("");
  }
  if (cmd.example) {
    lines.push(`*Example:*\n\`${cmd.example}\``);
    lines.push("");
  }
  if (cmd.sample_output) {
    lines.push(`*Sample Output:*\n\`\`\`\n${cmd.sample_output}\n\`\`\``);
    lines.push("");
  }
  if (cmd.privilege_level || cmd.configuration_mode) {
    lines.push(
      `*Privilege Level:* ${cmd.privilege_level ?? "N/A"}\n*Configuration Mode:* ${cmd.configuration_mode ?? "N/A"}`
    );
    lines.push("");
  }
  if (cmd.notes) {
    lines.push(`*Notes:*\n${cmd.notes}`);
    lines.push("");
  }

  return lines.join("\n").trim();
}

/** Compact one-line summary used in search result lists. */
export function formatCommandSummary(cmd: CiscoCommand): string {
  const primary = cmd.ios ?? cmd.ios_xe ?? cmd.ios_xr ?? cmd.nx_os ?? "";
  return `*${escapeMarkdown(cmd.title)}* — \`${primary}\``;
}

export function searchResultsKeyboard(commandIds: { id: string; title: string }[]): InlineKeyboardMarkup {
  return {
    inline_keyboard: commandIds.map((c) => [{ text: c.title, callback_data: `cmd:${c.id}` }]),
  };
}

export function commandDetailKeyboard(cmd: CiscoCommand): InlineKeyboardMarkup {
  const rows = [
    [
      { text: "⭐ Add to favorites", callback_data: `fav_add:${cmd.id}` },
      { text: "🔁 Convert", callback_data: `conv:${cmd.id}` },
    ],
  ];
  if (cmd.related_commands.length > 0) {
    rows.push(
      cmd.related_commands.slice(0, 3).map((id) => ({ text: `↪ ${id}`, callback_data: `cmd:${id}` }))
    );
  }
  return { inline_keyboard: rows };
}

export function categoriesKeyboard(categories: Category[]): InlineKeyboardMarkup {
  const rows: { text: string; callback_data: string }[][] = [];
  for (let i = 0; i < categories.length; i += 2) {
    const row = categories.slice(i, i + 2).map((c) => ({
      text: `${c.name} (${c.count})`,
      callback_data: `cat:${c.name}`,
    }));
    rows.push(row);
  }
  return { inline_keyboard: rows };
}

export function mainMenuKeyboard(): InlineKeyboardMarkup {
  return {
    inline_keyboard: [
      [
        { text: "🔍 Search", callback_data: "menu:search" },
        { text: "📂 Categories", callback_data: "menu:category" },
      ],
      [
        { text: "🎲 Random", callback_data: "menu:random" },
        { text: "⭐ Favorites", callback_data: "menu:favorites" },
      ],
      [
        { text: "🕓 History", callback_data: "menu:history" },
        { text: "ℹ️ About", callback_data: "menu:about" },
      ],
    ],
  };
}

export function convertKeyboard(cmd: CiscoCommand): InlineKeyboardMarkup {
  return { inline_keyboard: [[{ text: "⬅ Back", callback_data: `cmd:${cmd.id}` }]] };
}
